doc/INSTALL.md

```diff
@@ -305,6 +305,12 @@ need to include `testnet=1`
     ./lightningd/lightningd &
     ./cli/lightning-cli help
 
+
+To install the built binaries into your system, you'll need to run `make install`:
+
+    sudo PATH="/usr/local/opt:$PATH"  LIBRARY_PATH=/opt/homebrew/lib CPATH=/opt/homebrew/include make install
```

Comment:
Thanks for working on CoreLightning on Mac! Excited to see more mac people!

I believe this should just be a plain `make install` though.

For homebrew packages we can specify directory prefixes using ./configure --prefix which is The Correct Way(tm) to handle that.

We shouldn't need sudo either as the default install location (/usr/local/bin) should normally be writable.

The homebrew directories might not be but we should let homebrew handle those anyway and let it tell us where to put things with configure.

Do you remember a specific error this was trying to solve?

Reply:
@niftynei

Reply:
It was failing to install without the `sudo`, but using `sudo` meant that the exports we had done early were no longer in scope; adding the exports inline along with sudo allowed for the install command to succeed.

Agree that maybe there's a less pathological way to do it, but it worked on my* machine TM :shrug:

*someone else's

Reply:
If we're transitioning to using `homebrew --prefix` then maybe this is unnecessary but afaiu the docs haven't been updated to include that yet?

Reply:
Hmmm @islandbitcoin can you run make install without sudo and paste the error it gives?

Here's make install's output on my machine:

```
% xcodebuild -version
Xcode 14.0.1
Build version 14A400

% sw_vers
ProductName:	macOS
ProductVersion:	12.6
BuildVersion:	21G115

% make install
mkdir -p /usr/local/bin
mkdir -p /usr/local/libexec/c-lightning
mkdir -p /usr/local/libexec/c-lightning/plugins
mkdir -p /usr/local/share/man/man1
mkdir -p /usr/local/share/man/man5
mkdir -p /usr/local/share/man/man7
mkdir -p /usr/local/share/man/man8
mkdir -p /usr/local/share/doc/c-lightning
cp tools/hsmtool tools/lightning-hsmtool
cc channeld/channeld.c
ld lightningd/lightning_channeld
install cli/lightning-cli lightningd/lightningd tools/lightning-hsmtool /usr/local/bin
install lightningd/lightning_channeld lightningd/lightning_closingd lightningd/lightning_connectd lightningd/lightning_dualopend lightningd/lightning_gossipd lightningd/lightning_hsmd lightningd/lightning_onchaind lightningd/lightning_openingd lightningd/lightning_websocketd /usr/local/libexec/c-lightning
[ -z "plugins/autoclean plugins/chanbackup plugins/bcli plugins/commando plugins/fetchinvoice plugins/funder plugins/topology plugins/keysend plugins/offers plugins/pay plugins/txprepare plugins/spenderp plugins/bookkeeper" ] || install plugins/autoclean plugins/chanbackup plugins/bcli plugins/commando plugins/fetchinvoice plugins/funder plugins/topology plugins/keysend plugins/offers plugins/pay plugins/txprepare plugins/spenderp plugins/bookkeeper /usr/local/libexec/c-lightning/plugins
install -m 644 doc/lightning-cli.1 /usr/local/share/man/man1
install -m 644 doc/lightningd-config.5 /usr/local/share/man/man5
install -m 644 doc/lightning-addgossip.7 doc/lightning-autoclean-status.7 doc/lightning-batching.7 doc/lightning-bkpr-channelsapy.7 doc/lightning-bkpr-dumpincomecsv.7 doc/lightning-bkpr-inspect.7 doc/lightning-bkpr-listaccountevents.7 doc/lightning-bkpr-listbalances.7 doc/lightning-bkpr-listincome.7 doc/lightning-check.7 doc/lightning-checkmessage.7 doc/lightning-close.7 doc/lightning-connect.7 doc/lightning-commando.7 doc/lightning-commando-rune.7 doc/lightning-createonion.7 doc/lightning-createinvoice.7 doc/lightning-datastore.7 doc/lightning-decodepay.7 doc/lightning-decode.7 doc/lightning-deldatastore.7 doc/lightning-delexpiredinvoice.7 doc/lightning-delforward.7 doc/lightning-delinvoice.7 doc/lightning-delpay.7 doc/lightning-disableoffer.7 doc/lightning-disconnect.7 doc/lightning-emergencyrecover.7 doc/lightning-feerates.7 doc/lightning-fetchinvoice.7 doc/lightning-fundchannel.7 doc/lightning-fundchannel_start.7 doc/lightning-fundchannel_complete.7 doc/lightning-fundchannel_cancel.7 doc/lightning-funderupdate.7 doc/lightning-fundpsbt.7 doc/lightning-getroute.7 doc/lightning-invoice.7 doc/lightning-keysend.7 doc/lightning-listchannels.7 doc/lightning-listdatastore.7 doc/lightning-listforwards.7 doc/lightning-listfunds.7 doc/lightning-listhtlcs.7 doc/lightning-listinvoices.7 doc/lightning-listoffers.7 doc/lightning-listpays.7 doc/lightning-listpeers.7 doc/lightning-listsendpays.7 doc/lightning-makesecret.7 doc/lightning-multifundchannel.7 doc/lightning-multiwithdraw.7 doc/lightning-newaddr.7 doc/lightning-notifications.7 doc/lightning-offer.7 doc/lightning-offerout.7 doc/lightning-openchannel_abort.7 doc/lightning-openchannel_bump.7 doc/lightning-openchannel_init.7 doc/lightning-openchannel_signed.7 doc/lightning-openchannel_update.7 doc/lightning-pay.7 doc/lightning-parsefeerate.7 doc/lightning-plugin.7 doc/lightning-recoverchannel.7 doc/lightning-reserveinputs.7 doc/lightning-sendinvoice.7 doc/lightning-sendonion.7 doc/lightning-sendonionmessage.7 doc/lightning-sendpay.7 doc/lightning-setchannel.7 doc/lightning-setchannelfee.7 doc/lightning-sendcustommsg.7 doc/lightning-signmessage.7 doc/lightning-staticbackup.7 doc/lightning-txprepare.7 doc/lightning-txdiscard.7 doc/lightning-txsend.7 doc/lightning-unreserveinputs.7 doc/lightning-utxopsbt.7 doc/lightning-waitinvoice.7 doc/lightning-waitanyinvoice.7 doc/lightning-waitblockheight.7 doc/lightning-waitsendpay.7 doc/lightning-withdraw.7 doc/lightning-ping.7 doc/lightning-stop.7 doc/lightning-signpsbt.7 doc/lightning-sendpsbt.7 doc/lightning-getinfo.7 doc/lightning-listtransactions.7 doc/lightning-listnodes.7 doc/lightning-listconfigs.7 doc/lightning-help.7 doc/lightning-getlog.7 /usr/local/share/man/man7
install -m 644 doc/lightningd.8 doc/lightning-hsmtool.8 /usr/local/share/man/man8
install -m 644 README.md doc/INSTALL.md doc/HACKING.md LICENSE /usr/local/share/doc/c-lightning
```

Reply:
> 

$:lightning dread$ make install
mkdir -p /usr/local/bin
mkdir -p /usr/local/libexec/c-lightning
mkdir -p /usr/local/libexec/c-lightning/plugins
mkdir -p /usr/local/share/man/man1
mkdir -p /usr/local/share/man/man5
mkdir -p /usr/local/share/man/man7
mkdir -p /usr/local/share/man/man8
mkdir -p /usr/local/share/doc/c-lightning
ld: library not found for -lgmp
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [tools/headerversions] Error 1

Reply:
Can you try `brew install gmp` and run `make install` again?

Reply:
already installed. 

```
Running `brew update --auto-update`...
==> Auto-updated Homebrew!
==> Updated Homebrew from 38cb50489 to faa995022.
Updated 2 taps (homebrew/core and homebrew/cask).
==> New Formulae
postgresql@15
==> New Casks
eloquent                        genesis-plus                    prismlauncher                   sms-plus                        topaz-denoise-ai                topaz-gigapixel-ai              topaz-sharpen-ai                topaz-video-ai

You have 12 outdated formulae installed.
You can upgrade them with brew upgrade
or list them with brew outdated.

Warning: gmp 6.2.1_1 is already installed and up-to-date.
To reinstall 6.2.1_1, run:
  brew reinstall gmp
```

Reply:
Weeeiiiirddd 🤔

Reply:
maybe gmp was installed with sudo and thus isn't in scope?

Reply:
Yeah could be. If you run `ls -ld $(find /usr/local | grep "gmp[.][la]\+")` it should tell us

Reply:
Wait need to run that as sudo to be sure 😂 `sudo ls -ld $(find /usr/local | grep "gmp[.][la]\+")`

Reply:
> 


```
$ sudo ls -ld $(sudo find /usr/local | grep "gmp[.][la]\+")
drwxr-xr-x  109 dread  staff  3488 Oct 21 15:08 .
```

Reply:
So they're not there. Weird! What's the output of `brew list gmp`?

Here's the output of these commands on my machine:
```
% sudo ls -ld $(sudo find /usr/local | grep "gmp[.][la]\+")
-r--r--r--  1 ddustin  staff  854824 Nov 14  2020 /usr/local/Cellar/gmp/6.2.1_1/lib/libgmp.a
-rw-r--r--  1 ddustin  staff  904480 Nov 13  2014 /usr/local/lib/libgmp.a
-rwxr-xr-x  1 ddustin  staff     907 Nov 13  2014 /usr/local/lib/libgmp.la

% brew list gmp
/usr/local/Cellar/gmp/6.2.1_1/include/ (2 files)
/usr/local/Cellar/gmp/6.2.1_1/lib/libgmp.10.dylib
/usr/local/Cellar/gmp/6.2.1_1/lib/libgmpxx.4.dylib
/usr/local/Cellar/gmp/6.2.1_1/lib/pkgconfig/ (2 files)
/usr/local/Cellar/gmp/6.2.1_1/lib/ (4 other files)
/usr/local/Cellar/gmp/6.2.1_1/share/info/ (3 files)
```

Reply:
Just to confirm you have run `xcode-select --install` in the past right?

Reply:
> xcode-select --install

yup

Reply:
Also what's the output of `brew link gmp`

Reply:
> brew link gmp

```
brew link gmp
Warning: Already linked: /opt/homebrew/Cellar/gmp/6.2.1_1
To relink, run:
  brew unlink gmp && brew link gmp
```

Reply:
Something very weird happened to your gmp installation

Outputs from these commands would be helpful:
`brew -v unlink gmp`
`brew -v link --overwrite --dry-run gmp`
`brew -v link --dry-run gmp`

You can probably just force it by doing 
`brew -v unlink gmp`
followed by
`brew -v link --overwrite gmp`

At the end `ls -ld $(find /usr/local | grep "gmp[.][la]\+")` should return something like mine:
```
% ls -ld $(find /usr/local | grep "gmp[.][la]\+")
-r--r--r--  1 ddustin  admin  854760 Nov 14  2020 /usr/local/Cellar/gmp/6.2.1_1/lib/libgmp.a
lrwxr-xr-x  1 ddustin  staff      34 Oct 21 16:16 /usr/local/lib/libgmp.a -> ../Cellar/gmp/6.2.1_1/lib/libgmp.a
```

Reply:
> ls -ld $(find /usr/local | grep "gmp[.][la]\+")

```

MacMax:lightning dread$ brew -v unlink gmp
Unlinking /opt/homebrew/Cellar/gmp/6.2.1_1... 
rm /opt/homebrew/include/gmp.h
rm /opt/homebrew/include/gmpxx.h
rm /opt/homebrew/lib/libgmp.10.dylib
rm /opt/homebrew/lib/libgmp.a
rm /opt/homebrew/lib/libgmp.dylib
rm /opt/homebrew/lib/libgmpxx.4.dylib
rm /opt/homebrew/lib/libgmpxx.a
rm /opt/homebrew/lib/libgmpxx.dylib
rm /opt/homebrew/lib/pkgconfig/gmp.pc
rm /opt/homebrew/lib/pkgconfig/gmpxx.pc
uninfo /opt/homebrew/share/info/gmp.info
rm /opt/homebrew/share/info/gmp.info
rm /opt/homebrew/share/info/gmp.info-1
rm /opt/homebrew/share/info/gmp.info-2
13 symlinks removed.
MacMax:lightning dread$ brew -v link --overwrite --dry-run gmp
Would remove:
info /opt/homebrew/share/info/gmp.info
MacMax:lightning dread$ brew -v link --dry-run gmp
Would link:
/opt/homebrew/include/gmp.h
/opt/homebrew/include/gmpxx.h
/opt/homebrew/share/info/gmp.info
info /opt/homebrew/share/info/gmp.info
/opt/homebrew/share/info/gmp.info-1
/opt/homebrew/share/info/gmp.info-2
/opt/homebrew/lib/libgmp.10.dylib
/opt/homebrew/lib/libgmp.a
/opt/homebrew/lib/libgmp.dylib
/opt/homebrew/lib/libgmpxx.4.dylib
/opt/homebrew/lib/libgmpxx.a
/opt/homebrew/lib/libgmpxx.dylib
/opt/homebrew/lib/pkgconfig/gmp.pc
/opt/homebrew/lib/pkgconfig/gmpxx.pc
MacMax:lightning dread$ brew -v link --overwrite gmp
Linking /opt/homebrew/Cellar/gmp/6.2.1_1... 
ln -s ../Cellar/gmp/6.2.1_1/include/gmp.h gmp.h
ln -s ../Cellar/gmp/6.2.1_1/include/gmpxx.h gmpxx.h
ln -s ../../Cellar/gmp/6.2.1_1/share/info/gmp.info gmp.info
info /opt/homebrew/share/info/gmp.info
ln -s ../../Cellar/gmp/6.2.1_1/share/info/gmp.info-1 gmp.info-1
ln -s ../../Cellar/gmp/6.2.1_1/share/info/gmp.info-2 gmp.info-2
ln -s ../Cellar/gmp/6.2.1_1/lib/libgmp.10.dylib libgmp.10.dylib
ln -s ../Cellar/gmp/6.2.1_1/lib/libgmp.a libgmp.a
ln -s ../Cellar/gmp/6.2.1_1/lib/libgmp.dylib libgmp.dylib
ln -s ../Cellar/gmp/6.2.1_1/lib/libgmpxx.4.dylib libgmpxx.4.dylib
ln -s ../Cellar/gmp/6.2.1_1/lib/libgmpxx.a libgmpxx.a
ln -s ../Cellar/gmp/6.2.1_1/lib/libgmpxx.dylib libgmpxx.dylib
ln -s ../../Cellar/gmp/6.2.1_1/lib/pkgconfig/gmp.pc gmp.pc
ln -s ../../Cellar/gmp/6.2.1_1/lib/pkgconfig/gmpxx.pc gmpxx.pc
13 symlinks created.
MacMax:lightning dread$ ls -ld $(find /usr/local | grep "gmp[.][la]\+")
drwxr-xr-x  109 dread  staff  3488 Oct 21 15:22 .
```

Reply:
Ah! That explains it. You installed homebrew in a custom directory. From homebrew's docs:

> The standard script installs Homebrew to /usr/local so that **[you don’t need sudo](https://github.com/Homebrew/brew/blob/664d0c67d5947605c914c4c56ebcfaa80cb6eca0/docs/FAQ.md) when you brew install**

https://github.com/Homebrew/brew/blob/664d0c67d5947605c914c4c56ebcfaa80cb6eca0/docs/Installation.md#untar-anywhere

> However do yourself a favor and install to /usr/local. Some things may not build when installed elsewhere. One of the reasons Homebrew just works relative to the competition is **because** we recommend installing to /usr/local. *Pick another prefix at your peril!*

You're probably stuck using sudo because of how you installed homebrew. I don't think we should encourage others to do the same 🙊

If you get a free couple hours it might be worth removing homebrew completely and reinstalling it the normal/recommended way as this is probably causing you problems across the board 😬

Reply:
are you limiting the instructions to users that only install homebrew without sudo?

Reply:
We could add a special section for people with custom homebrew setups, not a crazy idea.

How did you set yours up to end up with custom homebrew directories?
Is there some common way / guide people are following to set them up non-standard like you did?

Reply:
doesn't brew come with MacOS? I'm pretty sure it was just there on my M1

Reply:
I've never seen in preinstalled but if there's some new thing where it is I'd love to see that.

Reply:
Oh woah, brew chooses the custom directory automatically for M1 chips!

> If Homebrew is not installed, there should be no Homebrew files in /usr/local (for macOS Intel) or /opt/homebrew (for Apple Silicon).
Source: https://mac.install.guide/homebrew/3.html

Yeah we should definitely support this. Seems crazy brew would break stuff on M1 like this 😳

Reply:
Brew discussion on it here: https://github.com/Homebrew/brew/issues/9177

Reply:
Can you post the output of brew doctor

Reply:
By the way, the `gmp` headers missing after `brew install` for me was caused by accidentally using rosetta instead of the native brew version (I effectively have an intel and an M1 homebrew installed, and it gets really confusing really quickly).

Reply:
What if we added a this as a section for M1 users.

Without an M1 mac I can't test this but seems good to have this here now for anyone getting stuck and perhaps we can revisit cleaning up the process down the line.